server/ai_mediaserver.go

```diff
@@ -621,14 +622,144 @@ func (ls *LivepeerServer) StartLiveVideo() http.Handler {
+
+		resp, err := processAIRequest(ctx, params, req)
+		if err != nil {
+			clog.Errorf(perOrchCtx, "Error processing AI Request: %s", err)
+			params.liveParams.stopPipeline(err)
+			perOrchCancel()
+			return
+		}
+
+		if err := startProcessing(perOrchCtx, params, resp); err != nil {
+			clog.Errorf(ctx, "Error starting processing: %s", err)
+			params.liveParams.stopPipeline(err)
+			perOrchCancel()
+			return
+		}
+
+		<-params.liveParams.processing
+		perOrchCancel()
+
+		if !orchSwapper.shouldSwap(ctx) {
+			break
+		}
+		clog.Infof(ctx, "Retrying stream with a different orchestrator")
+	}
+	params.liveParams.stopPipeline(fmt.Errorf("Done processing"))
+	params.liveParams.outputWriter.Close()
+}
+
+func newParams(params *liveRequestParams) *liveRequestParams {
+	return &liveRequestParams{
+		segmentReader:          params.segmentReader,
+		outputRTMPURL:          params.outputRTMPURL,
+		mediaMTXOutputRTMPURL:  params.mediaMTXOutputRTMPURL,
+		stream:                 params.stream,
+		paymentProcessInterval: params.paymentProcessInterval,
+		requestID:              params.requestID,
+		streamID:               params.streamID,
+		pipelineID:             params.pipelineID,
+		modelID:                params.modelID,
+		stopPipeline:           params.stopPipeline,
+		sendErrorEvent:         params.sendErrorEvent,
+		processing:             make(chan struct{}),
+		outputWriter:           params.outputWriter,
+		startTime:              time.Now(),
+	}
+}
+
+func startOutput(ctx context.Context, params aiRequestParams) *multiWriter {
+	var err error
+	rMediaMTX, wMediaMTX, err := os.Pipe()
+	if err != nil {
+		params.liveParams.stop()
+		return nil
+	}
+	r, w, err := os.Pipe()
+	if err != nil {
+		params.liveParams.stop()
+		return nil
+	}
+	mWriter := &multiWriter{ctx: ctx, writers: []io.Writer{w, wMediaMTX}}
+	// Studio Output ffmpeg process
+	if params.liveParams.outputRTMPURL != "" {
+		go ffmpegOutput(ctx, params.liveParams.outputRTMPURL, r, params)
+	}
+	go ffmpegOutput(ctx, params.liveParams.mediaMTXOutputRTMPURL, rMediaMTX, params)
+	return mWriter
+}
+
+func startProcessing(ctx context.Context, params aiRequestParams, res interface{}) error {
+	resp := res.(*worker.GenLiveVideoToVideoResponse)
+
+	host := params.liveParams.sess.Transcoder()
+	pub, err := common.AppendHostname(resp.JSON200.PublishUrl, host)
+	if err != nil {
+		return fmt.Errorf("invalid publish URL: %w", err)
+	}
+	sub, err := common.AppendHostname(resp.JSON200.SubscribeUrl, host)
+	if err != nil {
+		return fmt.Errorf("invalid subscribe URL: %w", err)
+	}
+	control, err := common.AppendHostname(*resp.JSON200.ControlUrl, host)
+	if err != nil {
+		return fmt.Errorf("invalid control URL: %w", err)
+	}
+	events, err := common.AppendHostname(*resp.JSON200.EventsUrl, host)
+	if err != nil {
+		return fmt.Errorf("invalid events URL: %w", err)
+	}
+	clog.V(common.VERBOSE).Infof(ctx, "pub %s sub %s control %s events %s", pub, sub, control, events)
+
+	onFirstSegment := func() {
```

Comment:
I am not sure why this is a callback tbh - I know its how the code was originally written, but I don't understand why it was done like this. All of the other events are dispatched in-line with the code, and if we are refactoring all this we can do the same there and simplify by removing this callback.

Reply:
Agree, I moved it inline in a dedicated [refactoring PR](https://github.com/livepeer/go-livepeer/pull/3565): https://github.com/livepeer/go-livepeer/pull/3565/commits/59477b7063264883d1ac0b0632d0f8713e7aa794